### PR TITLE
Windows Separation Remastered: Preferences Window

### PR DESCRIPTION
**Pull request by lolgear**
- [x] Preferences window controller has been encapsulated.
- [x] Preferences window controller has been moved to its own xib.

### CODE REVIEW

**Comment by lolgear** [comment]
@lucasderraugh Ready, I guess :) ( except layout, heh. I will add constraints a bit later. )

**Comment by lucasderraugh** [comment]
> @lucasderraugh Ready, I guess :) ( except layout, heh. I will add constraints a bit later. )

I'll wait until you think it's all done haha. Looks pretty good so far though.

**Comment by lolgear** [comment]
@lucasderraugh Ok, layout has been added.
It still contains warnings "prior OS X 10.11". Well, maybe it is time to set deployment target to something newer than 10.10? I guess that 10.11 is pretty good.

**Comment by lolgear** [comment]
@lucasderraugh If we can raise deployment target, all this layout can be easily done by `NSGridView`.
It is available since 10.12.

**Comment by lolgear** [comment]
@lucasderraugh Go-go-go-go-go :)

**Comment by lucasderraugh** [comment]
Sorry, I merged the text size changes in first as those were changes made based on a branch many years ago. If you get the chance to look over the conflicts we can merge these in as well. I think we can consider bumping later, but for now constraints will work just fine. We already moved to 10.10 this year to get over some dark mode hurtles with older releases. I think perhaps we can consider moving to 10.12 on the next release of macOS.

**Comment by lolgear** [comment]
@lucasderraugh 
Fixed.

**Comment by lolgear** [comment]
@lucasderraugh 
If everything is fine and you would like to merge it, I guess it could be release candidate.
But, of course, I would like to merge other PRs also, especially #573 ❤️

**Comment by lolgear** [comment]
@lucasderraugh 
Also #583 could enhance UX for most of us...

File: GitUpKit/Utilities/GILaunchServicesLocator.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//  GitUpKit (OSX)
+//
+//  Created by Dmitry Lobanov on 08.10.2019.
+//
+
+#import <Foundation/Foundation.h>
+NS_ASSUME_NONNULL_BEGIN
+extern NSString* const GIViewController_DiffTool;
+extern NSString* const GIViewController_MergeTool;
+extern NSString* const GIViewController_TerminalTool;
```
Comment (lucasderraugh) [inline GitUpKit/Utilities/GILaunchServicesLocator.h]: Why `GIViewController`?

File: GitUpKit/Utilities/GILaunchServicesLocator.m [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    return GIViewController_TerminalTool_iTerm_BundleIdentifier;
+  }
+  return nil;
+}
++ (NSString *)standardDefaultsKeyForDisplayName:(NSString *)displayName {
+  if ([displayName isEqualToString:GIViewController_TerminalTool_iTerm]) {
+    return GIViewController_TerminalTool_iTerm_Key;
+  }
+  return nil;
+}
```
Comment (lucasderraugh) [inline GitUpKit/Utilities/GILaunchServicesLocator.m]: Static analysis warning for returning nil when expecting non-nil return.

File: GitUp/Application/PreferencesWindowController.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+extern NSString* const PreferencesWindowController_Theme_Dark;
+
+@interface PreferencesThemeService : NSObject
++ (NSString *)selectedTheme;
++ (void)applySelectedTheme;
+@end
+
+@interface PreferencesWindowController : NSWindowController
+@property (nonatomic, copy) NSArray <NSString *>*channelTitles;
+@property (nonatomic, copy) NSArray <NSString *>*themesTitles;
```
Comment (lucasderraugh) [inline GitUp/Application/PreferencesWindowController.h]: Nit: Generics right next to type: `NSArray<` (no space)

File: GitUpKit/Utilities/GILaunchServicesLocator.h (hunk 1 above)
Comment (lolgear) [reply GitUpKit/Utilities/GILaunchServicesLocator.h]: @lucasderraugh Legacy, I guess.
`GIPreferences_` prefix is better?

```objective-c
// Change
extern NSString *const GIPreferences_DiffTool;
extern NSString *const GIPreferences_MergeTool;
extern NSString *const GIPreferences_TerminalTool;

// DiffTool and MergeTool
extern NSString *const GIPreferences_DiffMergeTool_FileMerge;
extern NSString *const GIPreferences_DiffMergeTool_Kaleidoscope;
extern NSString *const GIPreferences_DiffMergeTool_BeyondCompare;
extern NSString *const GIPreferences_DiffMergeTool_P4Merge;
extern NSString *const GIPreferences_DiffMergeTool_GitTool;
extern NSString *const GIPreferences_DiffMergeTool_DiffMerge;

// TerminalTool
extern NSString *const GIPreferences_TerminalTool_Terminal;
extern NSString *const GIPreferences_TerminalTool_iTerm;
```

File: GitUp/Application/PreferencesWindowController.m [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+- (void)showWindow:(id)sender {
+  // sync NSUserDefaults
+  NSString* theme = [NSUserDefaults.standardUserDefaults stringForKey:kUserDefaultsKey_Theme];
+  self.selectedTheme = theme;
+  NSString* channel = [NSUserDefaults.standardUserDefaults stringForKey:kUserDefaultsKey_ReleaseChannel];
+  self.selectedChannel = channel;
+  [self selectPreferencePane:nil];
+  [super showWindow:sender];
+}
```
[the remark this answers is not in the document]
Comment (lolgear) [reply GitUp/Application/PreferencesWindowController.m]: @lucasderraugh 
Heh, weird.

```objective-c
  [self selectPreferencePane:self]; // not nil sender
```
fixed it.

Maybe we can remove checks for `sender` here:

```objective-c
- (IBAction)selectPreferencePane:(id)sender {
  NSWindow *window = self.window;
  [_preferencesTabView selectTabViewItemWithIdentifier:_preferencesToolbar.selectedItemIdentifier];
  NSSize size = NSSizeFromString(_preferencesTabView.selectedTabViewItem.label);
  NSRect rect = [window contentRectForFrameRect:window.frame];
//  if (sender) {
    rect.origin.y += rect.size.height;
//  }
  rect.size.width = size.width;
  rect.size.height = size.height;
//  if (sender) {
    rect.origin.y -= rect.size.height;
//  }
  [window setFrame:[window frameRectForContentRect:rect] display:YES animate:(sender ? YES : NO)];
}
```

File: GitUp/Application/PreferencesWindowController.h (hunk 3 above)
Comment (lolgear) [reply GitUp/Application/PreferencesWindowController.h]:
```objective-c
@property (nonatomic, copy) NSArray<NSString *>*themesTitles;

// or

@property (nonatomic, copy) NSArray<NSString*>*themesTitles;

// or 

@property (nonatomic, copy) NSArray<NSString*> *themesTitles;
```

File: GitUpKit/Utilities/GILaunchServicesLocator.h (hunk 1 above)
Comment (lolgear) [reply GitUpKit/Utilities/GILaunchServicesLocator.h]: Also I would like to add suffix to distinguish `Keys` and `DisplayNames`.

```objective-c
extern NSString *const GIPreferences_DiffTool_Key;
extern NSString *const GIPreferences_DiffMergeTool_FileMerge_DisplayName;
```

File: GitUp/Application/PreferencesWindowController.h (hunk 3 above)
Comment (lucasderraugh) [reply GitUp/Application/PreferencesWindowController.h]: The style throughout the rest of the project is to have the type all in one blob and then have the identifier after a space so my vote would be `NSArray<NSString*>* propertyName`. Not how my personal projects would go, but it's not really my repo 😅.

File: GitUpKit/Utilities/GILaunchServicesLocator.h (hunk 1 above)
Comment (lucasderraugh) [reply GitUpKit/Utilities/GILaunchServicesLocator.h]: Ah, I see. I think `GIPreferences` works well. I think you can shorten up some of these names. If it's just being used as the localized string value, just `GIPreferences_FileMerge` would work fine if the string is just going to be "File Merge" anyways.

File: GitUp/Application/PreferencesWindowController.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucasderraugh) [reply GitUp/Application/PreferencesWindowController.m]: I'm not sure really in what cases there is or isn't a sender? I would assume there always is one. I just want the preference pane to never move itself down haha.

File: GitUp/Application/PreferencesWindowController.h (hunk 3 above)
Comment (lolgear) [reply GitUp/Application/PreferencesWindowController.h]: @lucasderraugh Fixed. :3

File: GitUpKit/Utilities/GILaunchServicesLocator.m (hunk 2 above)
Comment (lolgear) [reply GitUpKit/Utilities/GILaunchServicesLocator.m]: @lucasderraugh Fixed.

File: GitUpKit/Utilities/GILaunchServicesLocator.h (hunk 1 above)
Comment (lolgear) [reply GitUpKit/Utilities/GILaunchServicesLocator.h]: @lucasderraugh Fixed.

File: GitUp/Application/PreferencesWindowController.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (lolgear) [reply GitUp/Application/PreferencesWindowController.m]: @lucasderraugh Fixed.

**Comment by lolgear** [comment]
@lucasderraugh Go-go-go

**Comment by lolgear** [comment]
Ready?

**Pull request merged**